draft-irtf-cfrg-opaque.md

```diff
@@ -398,145 +398,331 @@ This specification defines one particular AKE based on 3DH; see {{instantiations
 
-                                record
-                      ------------------------->
+Steps:
+1. if server_public_key == nil || client_public_key == nil
+2.  raise NoPubKeyError
+3. if server_identity == nil
+4.  server_identity = server_public_key
+5. if client_identity == nil
+6.  client_identity = client_public_key
+8. Create CleartextCredentials cleartext_credentials with (server_public_key, client_public_key, server_identity, client_identity)
+9. Output cleartext_credentials
 ~~~
 
-{{registration-functions}} describes details of the functions referenced above.
+During protocol execution, these values can be stored in an implementation-specific `Credentials` object with names matching the values. For example, `Credentials.server_public_key` is the server's public key.
 
-Both client and server MUST validate the other party's public key before use.
-See {{validation}} for more details.
+## Envelope Structure {#envelope-structure}
 
-Upon completion, S stores C's credentials for later use. See {{credential-file}}
-for a recommended storage format.
+A client `Envelope` is constructed independently from the `EnvelopeMode`, but offers an `InnerEnvelope` entry that modes can use as they see fit.
+`Envelope` is constructed as follows:
+~~~
+struct {
+ EnvelopeMode mode;
+ opaque nonce[Nn];
+ opaque auth_tag[Nh];
+ InnerEnvelope inner_env;
+} Envelope;
+~~~
 
-## Credential Storage {#credential-storage}
+mode : The `EnvelopeMode` used for the `Envelope`.
 
-OPAQUE makes use of a structure `Envelope` to store client credentials.
-The `Envelope` structure embeds the following types of credentials:
+nonce : A unique nonce of length `Nn` used to protect this Envelope.
 
-- client_private_key: The encoded client private key for the AKE protocol.
-- server_public_key: The encoded server public key for the AKE protocol.
-- client_identity: The client identity. This is an application-specific value, e.g., an e-mail
-  address or normal account name.
-- server_identity: The server identity. This is typically a domain name, e.g., example.com.
-  See {{identities}} for information about this identity.
+auth_tag : Authentication tag protecting the contents of the envelope, covering `InnerEnvelope` and `CleartextCredentials`.
+
+inner_env : A mode dependent `InnerEnvelope` specifying at least the `EnvelopeMode`.
+
+The `CreateEnvelope` function allows a client to create its `Envelope` at registration.
+~~~
+CreateEnvelope(y, pks, creds)
+
+Parameters:
+- params, the MHF parameters established out of band
+- mode, the EnvelopeMode mode
+- Nx, the output size of the Extract function
+- Nn, the nonce length
+
+Input:
+- y, finalized OPRF output
+- creds, a Credentials structure
+- pks, the server AKE public key
+
+Output:
+- envelope, the client's `Envelope` structure
+- client_public_key, the client's AKE public key
+- export_key, an additional key
+
+Steps:
+1. envelope_nonce = random(Nn)
+2. prk = Extract(envelope_nonce, Harden(y, params))
```

Comment:
@kevinlewi's change removed the nonce from this step, so we can probably lift derivation of `prk` to the function that calls `CreateEnvelope`.